src/sp/it/pl/util/hotkey/Hotkeys.kt

```diff
@@ -0,0 +1,174 @@
+                GlobalScreen.addNativeKeyListener(keyListener)
+                GlobalScreen.registerNativeHook()
+                this.keyListener = keyListener
+            } catch (e: NativeHookException) {
+                logger.error("Failed to register global hotkeys", e)
+            }
+
+        }
+    }
+
+    fun stop() {
+        if (isRunning()) {
+            try {
+                logger.info { "Stopping global hotkeys" }
+                GlobalScreen.removeNativeKeyListener(keyListener)
+                keyListener = null
+                GlobalScreen.unregisterNativeHook()
+            } catch (e: NativeHookException) {
+                logger.error(e) { "Failed to unregister global hotkeys" }
+            }
+
+        }
+    }
+
+    fun register(action: Action, keys: String) {
+        val keyString = keys.substringAfterLast('+').trim().toLowerCase().replace('_', ' ')
+        val key = KeyCode.values().find { k -> k.getName().equals(keyString, true) }
+                ?: fail { "No KeyCode for ${action.keys}" }
+        register(
+                action,
+                key,
+                *sequenceOf(KeyCode.ALT, KeyCode.ALT_GRAPH, KeyCode.SHIFT, KeyCode.CONTROL, KeyCode.WINDOWS, KeyCode.COMMAND, KeyCode.META)
+                        .filter { k -> keys.contains(k.name, true) || (k==KeyCode.CONTROL && keys.contains("CTRL", true)) }
+                        .toList().toTypedArray()
+        )
+    }
+
+    fun register(action: Action, key: KeyCode, vararg modifiers: KeyCode) {
+        keyCombos[action.id] = KeyCombo(key, *modifiers)
+    }
+
+    fun unregister(action: Action) {
+        keyCombos.remove(action.id)
+    }
+
+    private inner class KeyCombo {
+        val key: KeyCode
+        val modifier: Int
+        var isPressed = false
+            private set(value) {
+                field = value
+            }
+
+        constructor(key: KeyCode, vararg modifiers: KeyCode) {
+            this.key = key
+            this.modifier = run {
+                var m = 0
+                if (KeyCode.SHIFT in modifiers) m += NativeInputEvent.SHIFT_L_MASK
+                if (KeyCode.CONTROL in modifiers) m += NativeInputEvent.CTRL_L_MASK
+                if (KeyCode.WINDOWS in modifiers || KeyCode.META in modifiers) m += NativeInputEvent.META_L_MASK
+                if (KeyCode.ALT in modifiers) m += NativeInputEvent.ALT_L_MASK
+                if (KeyCode.COMMAND in modifiers) m += NativeInputEvent.META_R_MASK
+                if (KeyCode.ALT_GRAPH in modifiers) m += NativeInputEvent.ALT_R_MASK
+                m
+            }
+        }
+
+        fun press(a: Action, e: NativeKeyEvent) {
+            val isPressedFirst = !isPressed
+            isPressed = true
+            if (a.isContinuous || isPressedFirst) {
+                // consume event (must be on the jNativeHook thread)
+                try {
+                    val f = NativeInputEvent::class.java.getDeclaredField("reserved")
```

Comment:
Why do we have to sue reflection here?

Reply:
Well, because jNativeHook is bugged or something and consuming the event has been officially removed at some point I think. It is still there, just accessible through reflection. The worst is, this must be accessed on the jnativehook thread for it to work! Which complicates the entire design of the class. But the feature is absolutely necessary and without it, we may as well ditch the library. Again. If you find an alternative to global hotkeys, im all ears. Im tired of the crashes and my cursor not working. I never asked for cursor monitoring in the 1st place. And the native code has clearly never dreamed async... We can literally hand Windows OS with just debugger in IntelliJ... Already had to restart my pc once!

Reply:
Ooof